src/tools/rbd_mirror/ImageDeleter.cc

```diff
@@ -0,0 +1,367 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2016 SUSE LINUX GmbH
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation.  See file COPYING.
+ *
+ */
+#include <boost/bind.hpp>
+#include <map>
+#include <set>
+#include <sstream>
+
+#include "include/rados/librados.hpp"
+#include "common/debug.h"
+#include "common/errno.h"
+#include "include/stringify.h"
+#include "librbd/internal.h"
+#include "librbd/ImageCtx.h"
+#include "librbd/ImageState.h"
+#include "librbd/Operations.h"
+#include "cls/rbd/cls_rbd_client.h"
+#include "cls/rbd/cls_rbd_types.h"
+#include "librbd/Utils.h"
+#include "ImageDeleter.h"
+
+#define dout_subsys ceph_subsys_rbd_mirror
+#undef dout_prefix
+#define dout_prefix *_dout << "rbd-mirror: ImageDeleter::" << __func__ << ": "
+
+using std::string;
+using std::map;
+using std::stringstream;
+
+using librados::IoCtx;
+using librbd::ImageCtx;
+using namespace librbd;
+
+namespace rbd {
+namespace mirror {
+
+ImageDeleter::ImageDeleter(string remote_cluster_name, RadosRef local_cluster)
+  : m_remote_name(remote_cluster_name),
```

Comment:
You're right. In my first implementation I would only delete an non-primary image if the primary images was really deleted (and not only disabled) and to know that I would have to contact the remote cluster. That is why ImageDeleter was tied with a remote cluster.
Now we just need one ImageDeleter.